base/system/services/database.c

```diff
@@ -617,35 +617,98 @@ ScmCreateOrReferenceServiceImage(PSERVICE pService)
     return dwError;
 }
 
+/*static*/ VOID
+ScmDereferenceOrTerminateServiceImage(PSERVICE pService)
+{
+    ASSERT(pService->lpImage);
+    ASSERT(pService->lpImage->dwImageRunCount > 0);
+
+    /* Decrement the run counter */
+    pService->lpImage->dwImageRunCount--;
```

Comment:
Considering you are playing with threads, and I don't see any locking, should this use the Interlocked apis?